### PR TITLE
Add Visibility states

### PR DESCRIPTION
**Pull request by Ewoutvans**
**Updated**
Colors:
- Cyan: New
- Yellow: Needs changes
- Red: Removed
schermafbeelding 2018-02-12 om 11 11 02

schermafbeelding 2018-02-12 om 11 14 59

schermafbeelding 2018-02-11 om 18 21 49

This introduces the states from #413 
Deleting & NeedsChanges will be in a separate PR

### CODE REVIEW

File: conf/evolutions/default/86.sql [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# --- !Ups
+ALTER TABLE projects ADD COLUMN visibility INT NOT NUll DEFAULT 1;
+UPDATE projects SET visibility = 5 WHERE is_visible = false;
+ALTER TABLE projects DROP COLUMN is_visible;
+
+# --- !Downs
+ALTER TaBLE projects ADD COLUMN is_visible BOOLEAN default TRUE;
```
Comment (phase) [inline conf/evolutions/default/86.sql]: typo in `TaBLE`

File: app/ore/project/ProjectTask.scala [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    val future = actions.collect(draftFilter.fn, ProjectSortingStrategies.Default, -1, 0)
+    val projects = this.models.await(future).get
+
+    val dayAgo = System.currentTimeMillis() - draftExpire
+
+    projects.foreach(project => {
+      Logger.info(s"Found project: ${project.ownerName}/${project.slug}")
+      val createdAt = project.createdAt.getOrElse(Timestamp.from(Instant.now())).getTime
+      if (createdAt < dayAgo) {
+        Logger.info(s"Project ${project.ownerName}/${project.slug} is still in Draft changing to Public")
```
Comment (phase) [inline app/ore/project/ProjectTask.scala]: `Change $owner/$slug visibility from Draft to Public` would be a little clearer.

File: app/controllers/Application.scala [hunk 3]
```diff
@@ -60,10 +60,15 @@ final class Application @Inject()(data: DataHelper,
...
-      val visibleFilter: ModelFilter[Project] = if (!canHideProjects)
-        ModelFilter[Project](_.isVisible)
+      var visibleFilter: ModelFilter[Project] = if (!canHideProjects)
+        ModelFilter[Project](_.visibility === VisibilityTypes.Public)
       else
         ModelFilter.Empty
+      if (this.users.current.isDefined) {
+        val currentUser = this.users.current.get
+        visibleFilter = visibleFilter +|| (ModelFilter[Project](_.userId === currentUser.id.get)
+            +&& ((ModelFilter[Project](_.visibility === VisibilityTypes.NeedsChanges) +|| ModelFilter[Project](_.visibility === VisibilityTypes.Draft))))
```
Comment (phase) [inline app/controllers/Application.scala]: Doesn't this leave out projects that the user owns that `NeedsApproval`? I think checking for `ModelFilter[Project](_.visibility =!= VisibilityTypes.SoftDelete)` would be cleaner.

**Review by phase — CHANGES_REQUESTED** [review]
This is looking awesome!

**Comment by dualspiral** [comment]
I'd like to suggest a change to the text you have in the red box you have on a draft page. I assume it's user facing - it's written in the third person so looks odd on a message directed at the author. Most users won't see that message because the project is hidden! I'd suggest something like the following might be better (calculating the time remaining before publish):

"This project is hidden for the next xx hours to allow you to add your project's documentation, after which your project will be published automatically. Once you have completed added documentation, you can choose to publish your project early by clicking the "Publish" button."

I'm open to that being rephrased, but that message needs to be author targeted, not third-person targetted.

**Comment by dualspiral** [comment]
...just realised, the draft page isn't going to be hidden, is it? I misread what was in the ore channels and whatnot.

Still, the message should be different for authors and users. If users see that, the first line of what you have is fine. Sorry for any confusion!

**Comment by Ewoutvans** [comment]
no draft is hidden, we changed it after some discussions
but yeah the messages need to change I forgot to do that

**Comment by mbax** [comment]
To avoid confusion, I'd like to see 'draft' renamed to 'new' or 'newproject'. That way, if we decide we want an actual drafting process we can do that later.

**Review by felixoi — APPROVED** [review]

**Pull request merged**